### ISSUE TITLE
[Feature request] whisper word level timestamps

### ISSUE THREAD

**Issue by jozefchutka** [xenova/transformers.js#68]
I am new to both transformers.js and whisper, so I am sorry for a lame question in advance.

I am trying to make [return_timestamps](https://huggingface.co/docs/transformers/main_classes/pipelines#transformers.AutomaticSpeechRecognitionPipeline.__call__) parameter work...

I managed to customize [script.js](https://github.com/xenova/transformers.js/blob/main/assets/js/scripts.js#L447) from [transformer.js demo](https://xenova.github.io/transformers.js/) locally and added `data.generation.return_timestamps = "char"`; around line ~447 inside GENERATE_BUTTON click handler in order to pass the parameter. With that change in place I am seeing timestamp appears as chunks (`result` var in [worker.js](https://github.com/xenova/transformers.js/blob/main/assets/js/worker.js#L40)):

```
{
    "text": " And so my fellow Americans ask not what your country can do for you ask what you can do for your country.",
    "chunks": [
        {
            "timestamp": [0,8],
            "text": " And so my fellow Americans ask not what your country can do for you"
        },
        {
            "timestamp": [8,11],
            "text": " ask what you can do for your country."
        }
    ]
}
```

however the chunks are not "char level" granular as expected following the [return_timestamps](https://huggingface.co/docs/transformers/main_classes/pipelines#transformers.AutomaticSpeechRecognitionPipeline.__call__) doc.

I am looking for ideas how to achieve char/word level timestamp granularity with transform.js and whisper. Do some models/tools need to be updated and/or rebuild?

**Comment by xenova** [comment]
Hi there 👋  Yes, you are correct, Transformers.js doesn't yet support char-level granularity.

For whisper (which is the only speech-to-text model currently supported), the model itself doesn't produce char-level timestamps. Instead, they output "timestamp tokens", which are placed between tokens . This means they are essentially treated as token boundaries (which act as word-level timestamps).

I will try add this functionality once more speech-to-text models are supported! Do you have any suggestions for what we should prioritize? By far, the most popular speech-to-text model seems to be [wav2vec2](https://huggingface.co/jonatasgrosman/wav2vec2-large-xlsr-53-english) (34.7M downloads this month)! So, we can look into adding that one soon.

**Comment by jozefchutka** [comment]
wav2vec2 seems popular, worth to consider adding for sure, I hope it is possible to provide base/small versions of its model suitable for browser runtime

I have also came across https://github.com/ggerganov/whisper.cpp which provides karaoke style / word level output. But I am not sure this is compatible tech.

Also it would be nice to extend `callback_function` with progress information or an extra `progress_callback` so one can track the progress of transcription

**Comment by xenova** [comment]
> I have also came across https://github.com/ggerganov/whisper.cpp and https://github.com/jianfch/stable-ts which provides karaoke style / word level output. But I am not sure this is compatible tech.

There are indeed various implementations which are able to produce word-level (or lower) timestamps for whisper, but they all are essentially "hacks" to add this functionality (since whisper was not trained to do that). As stated [here](https://github.com/linto-ai/whisper-timestamped#notes-on-other-approaches):
> An alternative approach, that does not require an additional model, is to look at the probabilities of timestamp tokens estimated by the Whisper model after each (sub)word token is predicted. It was implemented for instance in [whisper.cpp](https://github.com/ggerganov/whisper.cpp) and [stable-ts](https://github.com/jianfch/stable-ts). But this approach lacks of robustness, because Whisper models do not have been trained to output meaningful timestamps after each word. Whisper models tend to predict timestamps only after a certain number of words have been predicted (typically at the end of a sentence), and the probability distribution of timestamps outside this condition may be inaccurate. In practice, these methods can produce results that are totally out-of-sync on some periods of time (we observed that especially when there is jingle music). Also the timestamp precision of Whisper models tend to be rounded to 1 second (as in many video subtitles), which is too inaccurate for words, and reaching a better accuracy is tricky.

Moreover, this functionality isn't currently present in the HF python library, so, it would probably be quite strange for the JavaScript implementation to have it before the original python version haha.

Maybe one day though! 😇

**Comment by jozefchutka** [comment]
Is https://github.com/linto-ai/whisper-timestamped something you are looking into? Seems doing exactly what I am looknig for

**Comment by xenova** [comment]
> Is https://github.com/linto-ai/whisper-timestamped something you are looking into? Seems doing exactly what I am looknig for

Yep! (That's one of the things I linked above). But as I said, it would probably be best if you open up a feature request to the Python library (https://github.com/huggingface/transformers) before we add it here 😊

**Comment by jozefchutka** [comment]
I have requested a feature https://github.com/huggingface/transformers/issues/22590 ... I hope I am doing things properly, these are my very first interaction with HF and transformers

**Comment by jozefchutka** [comment]
Adding more tts suggestions:

- https://github.com/elevenlabs/elevenlabs-python
- https://github.com/suno-ai/bark
- https://github.com/coqui-ai/TTS

Not sure these are suitable for FE/transformers.js considering architecture or eventually performance.

**Comment by jozefchutka** [comment]
word level timestamps has been merged into transformers https://github.com/huggingface/transformers/pull/23205 

please consider adding support for this in transformers.js

**Comment by jozefchutka** [comment]
I have modified title/tag of this issue to covert it into a feature request

**Comment by xenova** [comment]
Just an update: I've got most of it working on this branch: https://github.com/xenova/transformers.js/tree/whisper-word-level-timestamps 🔥🔥🔥  I think the last thing for me to do is test, but should be ready soon! Also expect to see a demo with https://github.com/xenova/whisper-web 😄 

Python (left) vs. JavaScript (right)

![image](https://github.com/xenova/transformers.js/assets/26504141/77568daa-def7-470c-ae01-38c992524902)

<details>
<summary>surprise me</summary>
<br>

![image](https://github.com/xenova/transformers.js/assets/26504141/8ec87dcc-303d-461d-838c-adef920d446a)


</details>

**Comment by xenova** [comment]
This is now available in v2.4.0! 🚀 

Release notes: https://github.com/xenova/transformers.js/releases/tag/2.4.0
Example usage:
```js
import { pipeline } from '@xenova/transformers';

let url = 'https://huggingface.co/datasets/Xenova/transformers.js-docs/resolve/main/jfk.wav';
let transcriber = await pipeline('automatic-speech-recognition', 'Xenova/whisper-tiny.en', {
    revision: 'output_attentions',
});
let output = await transcriber(url, { return_timestamps: 'word' });
// {
//   "text": " And so my fellow Americans ask not what your country can do for you ask what you can do for your country.",
//   "chunks": [
//     { "text": " And", "timestamp": [0, 0.78] },
//     { "text": " so", "timestamp": [0.78, 1.06] },
//     { "text": " my", "timestamp": [1.06, 1.46] },
//     ...
//     { "text": " for", "timestamp": [9.72, 9.92] },
//     { "text": " your", "timestamp": [9.92, 10.22] },
//     { "text": " country.", "timestamp": [10.22, 13.5] }
//   ]
// }
```
